### PR TITLE
[FLINK-21080][runtime][checkpoint] Handle UnionListState with finished operators

### PR DESCRIPTION
**Pull request by gaoyunhaii**
## What is the purpose of the change

This PR avoids that for a checkpoint, if an operator has used `UnionListState`, we would expected to include the state of all the subtasks to avoid possible issues. We achieve this by abort checkpoints if we found that
1. Part of tasks of a vertex has turn into FINISHED, and remaining tasks have reported `UnionListState`. 
2. All of the tasks of a vertex are running, but some tasks have called operators.finish() and others are not. 

Then

1. If for a vertex it does not wait for the final checkpoint (due to the possible optimization), it might come across both case 1 and case 2. After all the tasks of this vertex are finished, the checkpoint would recover. 
2. If for a vertex it waits for the final checkpoint, then the finished tasks would be blocked before waiting the final checkpoints since checkpoints are aborted. Until all the subtasks reach this position, the checkpoint would be successful, then one checkpoint would unblock all these tasks. 

## Brief change log

- 7a73395b692c6e8fa5d1964ec0b0919c617be1f0 Unified the terminology for `finishedOnRestore` to distinguish with `operatorsFinished`. 
- 4af31ccbdab3f24476c7b887528a47a9edf5e9a9 The task reports if its have called operators' finished method in snapshot.
- 981d6981f9dacfd285cb992d630a165a7338c558 When finalize checkpoint, check if there are vertices that used UnionListState is partly finished, or have part of tasks report operators finished.
- aedb0fe1f96c4a7acdf5b5e49d45008354b3963e If notification of checkpoint complete loss for vertex that have used UnionListState and report operators' finished, cause task failover in case the task would hang forever to wait for the final checkpoint, but which would never succeed again.

## Verifying this change

This PR is verified by the added UT. 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **no**
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive): **no**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: **no**
  - The S3 file system connector: **no**

## Documentation

  - Does this pull request introduce a new feature?  **no**
  - If yes, how is the feature documented? **not applicable**

### CODE REVIEW

**Comment by dawidwys** [comment]
Why do we need fe17aca528dc6e65b6e05ea96779b99fcce19002? In the PR (https://github.com/dawidwys/flink/tree/flink-23408-wait) I'm implementing completion of any later checkpoint will finish such a task and thus I think we do not need to make the `notifyCheckpointComplete` any more reliable than it is now.

Also how is it related to `UnionListState`?

**Comment by gaoyunhaii** [comment]
Hi @dawidwys , it is mainly due to the following case:

1. Suppose we have a job vertex that have two subtasks and indeed need to wait for the final checkpoint. Then the first task called  `operator.finish()` and start to wait for the final checkpoint. Then in the following checkpoint, it will report `isOperatorsFinished = true`. 
2. Since the second task has not reached here, it will continue to report `isOperatorsFinished=false`. Then the checkpoint would be aborted as expected.
3. Later the second tasks called `operator.finish()` and start to wait for the final checkpoint. Then for the next checkpoint, it will also report `isOperatorsFinished = true`. By here it is as expected since we want to avoid the incomplete checkpoint by force the two tasks wait for the same final checkpoint. 
4. However, if the completion notification to the first task succeed, but the one to the second task fails, then the first task would finish waiting and exit. But the second task would continue to wait, what's more, the following checkpoints would always failed since the first task has exit. Then the second task would wait for the final checkpoint forever. 

Thus we might need to keep the notification reliable if it is send to a vertex who has used `UnionListState` and report `isOperatorsFinished = true` in this checkpoint~

**Comment by dawidwys** [comment]
Ok, I think I remember I talked with @pnowojski about it. The solution we had for it, was to add the information about the latest completed checkpoint to the `abort` checkpoint method. That way we can in a way implement retrying `notifyCheckpointComplete` via following `notifyCheckpointAbort`.  In our opinion, it is slightly simpler and complicates the protocol a bit less than the retry/fail and sending over the list of tasks that require a reliable notification. WDYT?

@pnowojski Do I remember it correctly?

**Comment by gaoyunhaii** [comment]
> In our opinion, it is slightly simpler and complicates the protocol a bit less than the retry/fail and sending over the list of tasks that require a reliable notification.

Very thanks for the suggestion! It also seems to me that the method would be simpler and with less changes. I'll first modify the implementation to the version following this option~

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointPlan.java [hunk 1]
```diff
@@ -83,4 +88,8 @@
     public List<ExecutionJobVertex> getFullyFinishedJobVertex() {
         return fullyFinishedJobVertex;
     }
+
+    public boolean isAllowsCheckpointAfterTasksFinished() {
```
[the remark this answers is not in the document]
Comment (gaoyunhaii) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointPlan.java:92]: When I modified the name it occurs to me I may not need this flag, I could skip the check as long as the `finishedTasks` list is empty. Thus I removed this flag~

**Comment by gaoyunhaii** [comment]
Hi @dawidwys I updated the PR according to the above discussion for also report the latest completed checkpoint in the abort notification~

**Review by dawidwys — COMMENTED** [review]
Thanks for the update. Is the reliable commit concept still necessary with the logic in `notifyCheckpointAborted`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java [hunk 2]
```diff
@@ -241,6 +263,10 @@ public boolean isAcknowledgedBy(ExecutionAttemptID executionAttemptId) {
         return !notYetAcknowledgedTasks.containsKey(executionAttemptId);
     }
 
+    public Set<JobVertexID> getVertexRequireReliableCommit() {
```
[the remark this answers is not in the document]
Comment (gaoyunhaii) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java:266]: Yes, indeed, sorry for the miss, I'll remove these codes~

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java [hunk 3]
```diff
@@ -368,6 +400,85 @@ public CompletedCheckpoint finalizeCheckpoint(
...
+            ExecutionJobVertex vertex = entry.getKey();
+
+            // If the vertex is partly finished, then it must not used UnionListState
+            // due to it passed the previous check.
+            if (partlyFinishedVertex.containsKey(vertex.getJobVertexId())) {
+                continue;
+            }
+
+            if (entry.getValue() != vertex.getParallelism() && hasUsedUnionListState(vertex)) {
+                throw new IllegalStateException(
```
Comment (dawidwys) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java:441]: nit: I think `IllegalStateException` is not the best choice here. After all it's not so much of an illegal state.

**Review by dawidwys — COMMENTED** [review]
Overall, lgtm. Could you please fix the compilation error and squash the commits?

**Comment by gaoyunhaii** [comment]
Very thanks @dawidwys for the review! I'll fix the checkstyle and squash the commits~

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java (hunk 3 above)
Comment (gaoyunhaii) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java:441]: I'll change it to the `FlinkRuntimeException`~

**Pull request closed without merging**